masonry/src/passes/layout.rs

```diff
@@ -0,0 +1,217 @@
+//! Most of the logic for this pass happens in [`Widget::layout`] implementations.
+
+use tracing::{info_span, trace};
+use vello::kurbo::{Point, Rect, Size};
+
+use crate::render_root::RenderRoot;
+use crate::widget::WidgetState;
+use crate::{BoxConstraints, LayoutCtx, Widget, WidgetPod};
+
+// TODO - Replace with contains_rect once new Kurbo version is released.
+// See https://github.com/linebender/kurbo/pull/347
+/// Return `true` if all of `smaller` is within `larger`.
+fn rect_contains(larger: &Rect, smaller: &Rect) -> bool {
+    smaller.x0 >= larger.x0
+        && smaller.x1 <= larger.x1
+        && smaller.y0 >= larger.y0
+        && smaller.y1 <= larger.y1
+}
+
+// Returns "true" if the Widget's layout method was called, in which case debug checks
+// need to be run. (See 'called_widget' in WidgetPod::call_widget_method_with_checks)
+pub(crate) fn run_layout_inner<W: Widget>(
+    parent_ctx: &mut LayoutCtx<'_>,
+    pod: &mut WidgetPod<W>,
+    bc: &BoxConstraints,
+) -> bool {
+    let id = pod.id().to_raw();
+    let widget_mut = parent_ctx
+        .widget_children
+        .get_child_mut(id)
+        .expect("WidgetPod: inner widget not found in widget tree");
+    let mut state_mut = parent_ctx
+        .widget_state_children
+        .get_child_mut(id)
+        .expect("WidgetPod: inner widget not found in widget tree");
+    let widget = widget_mut.item;
+    let state = state_mut.item;
+
+    if state.is_stashed {
+        debug_panic!(
+            "Error in '{}' #{}: trying to compute layout of stashed widget.",
+            widget.short_type_name(),
+            id,
+        );
+        state.size = Size::ZERO;
+        return false;
+    }
+
+    // TODO - Not everything that has been re-laid out needs to be repainted.
+    state.needs_paint = true;
+    state.needs_compose = true;
+    state.needs_accessibility = true;
+    state.request_paint = true;
+    state.request_compose = true;
+    state.request_accessibility = true;
+
+    bc.debug_check(widget.short_type_name());
+    trace!("Computing layout with constraints {:?}", bc);
+
+    state.local_paint_rect = Rect::ZERO;
+
+    let new_size = {
+        let mut inner_ctx = LayoutCtx {
+            widget_state: state,
+            widget_state_children: state_mut.children.reborrow_mut(),
+            widget_children: widget_mut.children,
+            global_state: parent_ctx.global_state,
+            mouse_pos: parent_ctx.mouse_pos,
+        };
+
+        // TODO - If constraints are the same and request_layout isn't set,
+        // skip calling layout
+        inner_ctx.widget_state.request_layout = false;
+        widget.layout(&mut inner_ctx, bc)
+    };
+    if state.request_layout {
+        debug_panic!(
+            "Error in '{}' #{}: layout request flag was set during layout pass",
+            widget.short_type_name(),
+            id,
+        );
+    }
+
+    state.needs_layout = false;
```

Comment:
There's a little bit of knowledge encoded in this line, so it could be worth having a comment here.

Not going to block on that, though